### PR TITLE
[CI] Add step to verify `symfony/deprecation-contracts` requirements

### PR DESCRIPTION
**Pull request by fancyweb**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | -
| License       | MIT

Screenshot from 2023-10-19 19-23-01

### CODE REVIEW

File: .github/workflows/package-tests.yml [hunk 1]
```diff
@@ -100,4 +100,41 @@ jobs:
...
+          ok=0
+
+          json='${{ steps.find-packages.outputs.packages }}'
+          for package in $(echo "${json}" | jq -r '.[] | @base64'); do
+              _jq() {
+                echo ${package} | base64 --decode | jq -r ${1}
+              }
+
+              NAME=$(_jq '.name')
+              if [[ $NAME = 'symfony/deprecation-contracts' || $NAME = 'symfony/contracts' ]]; then
```
Comment (fancyweb) [inline .github/workflows/package-tests.yml:116]: Those two don't require it but they always have it

File: .github/workflows/package-tests.yml [hunk 2]
```diff
@@ -100,4 +100,41 @@ jobs:
...
+                continue
+              fi
+
+              echo "::group::$NAME"
+              DIR=$(_jq '.directory')
+              localExit=0
+              grep -rq 'trigger_deprecation(' --include=*.php --exclude-dir=Tests/ --exclude-dir=Bridge/ $DIR
```
Comment (fancyweb) [inline .github/workflows/package-tests.yml:123]: We exclude tests and component bridges that are in the root component code but that are released as separate packages

File: .github/workflows/package-tests.yml [hunk 3]
```diff
@@ -100,4 +100,41 @@ jobs:
...
+              triggersDeprecation=$?
+              if [[ $triggersDeprecation -eq 0 && $(_jq '.requires_deprecation_contracts') == false ]]; then
```
Comment (fancyweb) [inline .github/workflows/package-tests.yml:125]: if there is at least one trigger_deprecation( match and the dep is not required -> fail

File: .github/workflows/package-tests.yml [hunk 4]
```diff
@@ -100,4 +100,41 @@ jobs:
...
+                localExit=1
+              elif [[ $triggersDeprecation -eq 1 && $(_jq '.requires_deprecation_contracts') == true ]]; then
```
Comment (fancyweb) [inline .github/workflows/package-tests.yml:128]: if there no trigger_deprecation( match and the dep is required -> fail

**Comment by fancyweb** [comment]
There will probably be false positives one day but it currently works for every maintained branch.

File: .github/workflows/package-tests.yml [hunk 5]
```diff
@@ -100,4 +100,41 @@ jobs:
...
+                localExit=1
+              elif [[ $triggersDeprecation -ne 0 && $triggersDeprecation -ne 1 ]]; then
+                echo "::error::grep failed"
+                exit 2
+              fi
+
+              ok=$(( $localExit || $ok ))
+              echo ::endgroup::
+              if [ $localExit -ne 0 ]; then
+                echo "::error::$NAME failed"
```
Comment (stof) [inline .github/workflows/package-tests.yml]: I would rather echo separate messages in the 2 failure cases, which allow to be more precise about why they fail (missing requirement vs useless requirement)

**Comment by OskarStark** [comment]
Maybe a dumb question, but couldn't we use composer-unused for this task? 

cc @icanhazstring

**Comment by stof** [comment]
@OskarStark not for the part about missing the necessary requirement (which is the most important part of this PR)

**Review by stof — APPROVED** [review]

**Comment by OskarStark** [comment]
Ah yes, you are right the other way around would be [ComposerRequireChecker ](https://github.com/maglnet/ComposerRequireChecker)

**Pull request merged**